<h1>Inspect: A Proteomics Search Toolkit</h1>
Copyright 2007, The Regents of the University of California
<h3>Table of Contents</h1>
<li><a href="index.html">Overview</a>
<li><a href="Copyright.html">Copyright information</a>
<li><a href="Installation.html">Installation</a>
<li><a href="Database.html">Database</a>
<li><a href="Searching.html">Searching</a>
<li><a href="Analysis.html">Analysis</a>
<li><a href="InspectTutorial.pdf">Basic Tutorial</a>
<li><a href="InspectAdvancedTutorial.pdf">Advanced Tutorial</a>
<li><a href="UnrestrictedSearchTutorial.pdf">Unrestricted Search Tutorial</a>

<hr>
<h1>Searching</h1>

To run a search, you first create an inspect <b>input file</b>.  The input file is text file that
tells Inspect what to do.  Each line of the
input file has the form [COMMAND],[VALUE].  For example, one line might be "spectra,spec18.dta",
where the command is "spectra" and the value is "spec18.dta".  Inspect ignores blank lines.  You can
include comments by starting lines with a hash character (#).  Here is an example
of what an input file might look like:
<pre>
spectra,Fraction01.mzxml
instrument,ESI-ION-TRAP
protease,Trypsin
DB,TestDatabase.trie
# Protecting group on cysteine:
mod,57,C,fix
</pre>
<br>Here are the
available input file commands.  Those you are most likely to set are listed first.  The only required
commands are one or more "spectra" commands, and either "db" or "SequenceFile".  Commands are
case-insensitive (type "Spectra" or "spectra", it doesn't matter).  Values are case-insensitive with
the exception (on Linux) of filenames.  If Inspect doesn't understand a command, it will print a
warning and ignore it.
<br><br>
<li><b>spectra,[FILENAME]</b> - Specifies a spectrum file to search.  You can specify the name of a
directory to search every file in that directory (non-recursively).<br>
Preferred file formats: .mzXML and .mgf <br>
Other accepted file formats: .mzData, .ms2 .dta.
Note that multiple spectra in a single .dta file are <b>not</b> supported.
<br>
<li><b>db,[FILENAME]</b> - Specifies the name of a database (.trie file) to search.  The .trie file
contains one or more protein sequences delimited by asterisks, with no whitespace or other data.
Use PrepDB.py (see <a href="Database.html">Databases</a> to prepare a database.  You should specify
at least one database.  You may specify several databases; if so, each database will be searched in turn.
<li><b>SequenceFile,[FILENAME]</b> - Specifies the name of a FASTA-format protein database to search.  If
you plan to search a large database, it is more efficient to preprocess it using PrepDB.py and use the "db"
command instead.  You can specify at most one SequenceFile.
<br>
<li><b>protease,[NAME]</b> - Specifies the name of a protease.  "Trypsin", "None", and "Chymotrypsin" are
the available values.  If tryptic digest is specified, then matches with non-tryptic termini are penalized.
<br>
<li><b>mod,[MASS],[RESIDUES],[TYPE],[NAME]</b> - Specifies an amino acid modification.  The delta mass
(in daltons) and affected amino acids are required.  The first four characters of the name should be
unique.  Valid values for "type" are "fix", "cterminal", "nterminal", and "opt" (the default).  For a guide
to various known modification types, consult the following databases:
<li> <a href="http://www.abrf.org/index.cfm/dm.home">ABRF mass delta reference</a>
<li> <a href="http://www.unimod.org">UNIMOD database</a>
<li>RESID database of modifications
Examples:
<br><tt>mod,+57,C,fix</tt> - Most searches should include this line.  It reflects the addition of CAM
(carbamidomethylation, done by adding iodoacetamide) which prevents cysteines from forming disulfide bonds.
<br><tt>mod,80,STY,opt,phosphorylation</tt>
<br><tt>mod,16,M</tt> (Oxidation of methionine, seen in many samples)
<br><tt>mod,43,*,nterminal</tt> (N-terminal carbamylation, common if sample is treated with urea)
<br>
<b>Important note:</b> When searching for phosphorylation sites, use a modification with the name "phosphorylation".
This lets Inspect know that it should use its model of phosphopeptide fragmentation
when generating tags and scoring matches.  (Phosphorylation of serine dramatically affects fragmentation, so
modeling it as simply an 80Da offset is typically <b>not</b> sufficient to detect sites with high sensitivity)
<li><b>Mods,[COUNT]</b> - Number of PTMs permitted in a single peptide.  Set this to 1 (or higher) if you
               specify PTMs to search for.
<li><b>Unrestrictive,[FLAG]</b> - If FLAG is 1, use the MS-Alignment algorithm to perform an <b>unrestrictive</b>
search (allowing arbitrary modification masses).  Running an unrestrictive search with one mod per peptide is slower than the
normal (tag-based) search; running time is approximately 1 second per spectrum per megabyte of database.  Running an unrestrictive search
with two mods is significantly slower.  We recommend performing unrestrictive searches against a small database, containing proteins
output by an earlier search.  (The "Summary.py" script can be used to generate a second-pass database
from initial search results; see <a href="Analysis">Analysis</a>)
<li><b>MaxPTMSize,[SIZE]</b> - For blind search, specifies the maximum modification size (in Da) to consider.
Defaults to 250.  Larger values require more time to search.
<li><b>PMTolerance,[MASS]</b> - Specifies the parent mass tolerance, in Daltons.  A candidate's
flanking mass can differ from the tag's flanking mass by no more than ths amount.  Default value
is 2.5.  Note that secondary ions are often selected for fragmentation, so parent mass errors near
1.0Da or -1.0Da are not uncommon in typical datasets, even on FT machines.
<li><b>ParentPPM,[MASS]</b> - Specifies a parent mass tolerance, in parts per million.  Alternative to PMTolerance.
<li><b>IonTolerance,[MASS]</b> - Error tolerance for how far ion fragments (b and y peaks) can be
    shifted from their expected masses.  Default is 0.5.  Higher values produce a more sensitive but much slower search.
<li><b>PeakPPM,[MASS]</b> - Specifies a fragment mass tolerance, in parts per million.  Alternative to IonTolerance.
<li><b>MultiCharge,[FLAG]</b> - If set to true, attempt to guess the precursor charge and mass, and consider
multiple charge states if feasible.
<li><b>Instrument,[TYPE]</b> - Options are ESI-ION-TRAP (default), QTOF, and FT-Hybrid.  If set to ESI-ION-TRAP,
Inspect attempts to correct the parent mass.  If set to QTOF, Inspect uses a fragmentation model trained on
QTOF data.  (QTOF data typically features a stronger y ladder and weaker b ladder than other spectra).
<li><b>RequiredMod,[NAME]</b> - The specified modification MUST be found somewhere on the peptide.
<li><b>TagCount,[COUNT]</b> - Number of tags to generate
<li><b>TagLength,[LENGTH]</b> - Length of peptide sequence tags.  Defaults to 3.  Accepted values are 1 through 6.
<li><b>RequireTermini,[COUNT]</b> - If set to 1 or 2, require 1 or 2 valid proteolytic termini.  Deprecated, because
    the scoring model already incorporates the number of valid (tryptic) termini.
<h3>Non-standard options:</h3>
<b>TagsOnly</b> - Tags are generated and written to the specified output file.  No search is performed.

<hr>
<h2>Command-line arguments</h2>
Inspect features a few command-line options.  Most options are specified in an <b>input file</b>, rather
than on the command-line.   The
command-line options are:
    <li> <b>-i</b> Input file name.  Defaults to "Input.txt"
    <li> <b>-o</b> Output file name.  Defaults to "Inspect.txt"
    <li> <b>-e</b> Error file name.  Defaults to "Inspect.err".
    <li> <b>-r</b> The resource directory.  Defaults to the current working directory.  The resource directory
is where Inspect searches for its resource files such as AminoAcidMasses.txt.
<br><br>
Sample usage:<br>
On Windows: <b>Inspect -i TripureIn.txt -o TripureOut.txt</b><br>
On Linux: <b>./inspect -i TripureIn.txt -o TripureOut.txt</b><br>
<h3>Error Reporting</h3>
If Inspect encounters a problem - such as a spectrum file with garbled format, or
running out of memory - it reports the problem to the error file.  One error (or warning)
is reported per line of the file, and each error/warning type has an ID, to make them
easier to parse.  If no error file is left behind after a run, then there were no errors - this
is a good thing!
<br><br>Here is a sample error message, where I gave inspect an incorrect file name:<br>
<tt>[E0008] .\ParseInput.c:725:Unable to open requested file '.\Database\TestDatbaase.trie'</tt>
<br>
And here is a sample warning message, where - on a small search - Inspect was not able to re-fit the p-value distribution:<br>
<tt>{W0010} .\PValue.c:396:Few spectra were searched; not recalibrating the p-value curve.</tt><br>
